contrib/hooks/lib.py

```diff
@@ -0,0 +1,139 @@
+"""
+Utility functions for hook scripts.
+"""
+
+import sys
+import os
+import subprocess
+import socket
+
+from subprocess import CalledProcessError, TimeoutExpired
+from typing import Tuple, Optional
+
+def complain(message):
+    sys.stderr.flush()
+    sys.stderr.write(message)
+    sys.stderr.write('\n')
+    sys.stderr.flush()
+
+def announce(message):
+    sys.stderr.flush()
+    sys.stderr.write(message)
+    sys.stderr.write('\n')
+    sys.stderr.flush()
+
+cached_hostname = None
+def get_hostname() -> str:
+    global cached_hostname
+    if cached_hostname is None:
+        try:
+            cached_hostname = socket.getfqdn()
+        except:
+            cached_hostname = 'localhost'
+    return cached_hostname
+
+def file_link(path: str, text: Optional[str] = None) -> str:
+    
+    # Determine link protocol to use
+    if 'SSH_TTY' in os.environ:
+        protocol = 'sftp'
+    else:
+        protocol = 'file'
+    
+    realpath = os.path.realpath(path)
+    if not text:
+        text = path    
+        
+    return f'\033]8;;{protocol}://{get_hostname()}{realpath}\033\\{text}\033]8;;\033\\'
```

Comment:
Could there be a comment here about the purpose of the `\033]8;;` / `\033]8;;\033\\` prefix/suffixes?

I found this, so assume it is affecting/minimizing terminal size for some reason?  https://stackoverflow.com/questions/68931529/printf-0338widthheightt-works-for-a-specific-height-and-width-but-is-t

Reply:
This is making [hyperlinks in the terminal with OSC 8](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).

I've rearranged this function to make it less magic and actually explain what it is up to and how the sequences fit together.